File: bindings/kotlin/examples/GasStation.kt
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import iota_sdk.*
import kotlinx.coroutines.runBlocking

fun main() = runBlocking {
    try {
        val client = GraphQlClient.newLocalnet()
        var gasStationUrl = "http://0.0.0.0:9527"
        var gasStationAuthToken = "test"
        var keypair = Ed25519PrivateKey.random()
        var sender = keypair.publicKey().deriveAddress()
        var signer = TransactionSigner.fromEd25519(keypair)

        val builder = TransactionBuilder(sender).withClient(client)

        builder.moveCall(
            Address.std(),
            Identifier("u64"),
            Identifier("sqrt"),
            listOf(PtbArgument.u64(64uL)),
        )

        builder.gasStationSponsor(
            gasStationUrl,
            headers = mapOf("Authorization" to listOf("Bearer $gasStationAuthToken")),
        )

        val res = builder.execute(signer)

        println("$res")

        println("Sponsored transaction was successful!")
    } catch (e: Exception) {
        e.printStackTrace()
        kotlin.system.exitProcess(1)
    }
}

File: bindings/swift/examples/GasStation.swift
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import Foundation
import IotaSDK

@main
struct GasStationExample {
  static func main() async {
    do {
      let client = GraphQlClient.newLocalnet()
      let gasStationUrl = "http://0.0.0.0:9527"
      let gasStationAuthToken = "test"
      let keypair = Ed25519PrivateKey.random()
      let sender = keypair.publicKey().deriveAddress()
      let signer = TransactionSigner.fromEd25519(key: keypair)

      let builder = client.transactionBuilder(sender: sender)

      _ = try builder.moveCall(
        package: Address.std(),
        module: Identifier(identifier: "u64"),
        function: Identifier(identifier: "sqrt"),
        arguments: [PtbArgument.u64(value: 64)]
      )

      _ = builder.gasStationSponsor(
        url: gasStationUrl,
        headers: ["Authorization": ["Bearer \(gasStationAuthToken)"]])

      let res = try await builder.execute(signer: signer)

      print(res)

      print("Sponsored transaction was successful!")
    } catch {
      print("Error: \(error)")
      exit(1)
    }
  }
}
